### PR TITLE
Fix limit setting and add ServerImp tests (RIPD-1463,1458):

### PR DESCRIPTION
**Pull request by mellery451**
Add more test coverage for ServerHandlerImp.cpp. Ensure limit parameter
is propagated from parsed object to in-memory config.

Release Notes
-------------

This fixes a bug whereby the limit parameter on a port configuration was
ignored.

### CODE REVIEW

File: src/test/server/Server_test.cpp [hunk 1]
```diff
@@ -367,10 +370,98 @@ class Server_test : public beast::unit_test::suite
     }
 
     void
+    testBadConfig ()
```
Comment (wilsonianb) [inline src/test/server/Server_test.cpp]: This test is outputting (expected) error messages. You could do something like [this](https://github.com/ripple/rippled/blame/b7e2a3bd5f0d5156893b1797fd2fb51ef5a33e0e/src/test/core/TerminateHandler_test.cpp#L36) to redirect cout in order to make the test less verbose.

Comment (wilsonianb) [inline src/test/server/Server_test.cpp]: I think that capturing the output would also allow you to check that exceptions are being thrown were you expect.

Comment (mellery451) [inline src/test/server/Server_test.cpp]: 👍  good suggestion

File: src/test/server/ServerStatus_test.cpp [hunk 2]
```diff
@@ -422,117 +516,346 @@ class ServerStatus_test :
...
+        doHTTPRequest(env, yield, false, resp, ec, to_string(jr), auth);
+        BEAST_EXPECT(resp.status == 200);
+        BEAST_EXPECT(! resp.body.empty());
+    }
 
-        Json::Value jrr;
+    void
+    testLimit(boost::asio::yield_context& yield)
+    {
+        testcase ("Server with connection limit");
```
Comment (wilsonianb) [inline src/test/server/ServerStatus_test.cpp]: Do you know if this `limit` option is deprecated? I'm wondering if a description should be added to the [example config](https://github.com/ripple/rippled/blob/develop/doc/rippled-example.cfg#L141-L144).

File: src/test/server/ServerStatus_test.cpp [hunk 3]
```diff
@@ -422,117 +516,346 @@ class ServerStatus_test :
...
+            clients.emplace_back(
+                std::make_pair(ip::tcp::socket {ios}, beast::streambuf {}));
+            async_connect(clients.back().first, it, yield[ec]);
+            BEAST_EXPECT(! ec);
+            async_write(
+                clients.back().first,
+                makeHTTPRequest(ip, port, to_string(jr), {}),
+                yield[ec]);
+            BEAST_EXPECT(! ec);
+            connectionCount++;
```
Comment (wilsonianb) [inline src/test/server/ServerStatus_test.cpp]: Nit
--> `++connectionCount`
also for `readCount` below

File: src/test/server/ServerStatus_test.cpp [hunk 4]
```diff
@@ -422,117 +516,346 @@ class ServerStatus_test :
...
+    }
+
+    void
+    testWSHandoff(boost::asio::yield_context& yield)
+    {
+        testcase ("Connection with WS handoff");
+
+        using namespace test::jtx;
+        Env env {*this, envconfig([](std::unique_ptr<Config> cfg) {
+            (*cfg)["port_rpc"].set("protocol","https,wss");
```
Comment (mellery451) [inline src/test/server/ServerStatus_test.cpp]: 👍

File: src/test/server/ServerStatus_test.cpp (hunk 2 above)
Comment (mellery451) [inline src/test/server/ServerStatus_test.cpp]: it looks like the `limit` param was added in 2016, so I doubt that it is deprecated. I will add it to the config doc.

File: doc/rippled-example.cfg [hunk 5]
```diff
@@ -172,6 +172,13 @@
 #       NOTE    If no ports support the peer protocol, rippled cannot
 #               receive incoming peer connections or become a superpeer.
 #
+#   limit = <number>
+#
+#       Optional. An integer value that will limit the number of connected
+#       clients that the port will accept. Once the limit is reached, new
+#       connections will be refused until other clients disconnect.
+#       Omit or set to 0 to allow unlimited numbers of clients.
```
Comment (wilsonianb) [inline doc/rippled-example.cfg]: Now that you mention it :smile:, could you update `testLimit` to check that a `limit` of 0 doesn't mean allow zero connections?

Comment (mellery451) [inline doc/rippled-example.cfg]: good idea - done.

File: src/test/server/Server_test.cpp [hunk 6]
```diff
@@ -366,11 +369,164 @@ class Server_test : public beast::unit_test::suite
         }
     }
 
+    class CaptureStreams
+    {
+    public:
+        CaptureStreams (std::string* save_messages)
+        : save_to (save_messages)
```
Comment (nbougalis) [inline src/test/server/Server_test.cpp]: Micronit: I'd indent these one stop.

File: src/beast/extras/beast/unit_test/reporter.hpp [hunk 7]
```diff
@@ -283,6 +283,7 @@ reporter<_>::
 on_log(std::string const& s)
 {
     os_ << s;
+    os_.flush();
```
Comment (mellery451) [inline src/beast/extras/beast/unit_test/reporter.hpp]: ah - good point. I'll remove this here and figure out the upstream strategy

Comment (ximinez) [inline src/beast/extras/beast/unit_test/reporter.hpp]: I don't think there's anything wrong with a local patch while we wait for upstream, especially if this is the only thing that fixes our issue.

**Comment by mellery451** [comment]
@bachase and @ximinez - I would appreciate your review of this latest commit: https://github.com/ripple/rippled/pull/2097/commits/ff13bd032c9187cf68ffdd3a02b1756fdbf6deda
 (resolves the issues with `cout` for validation).

File: src/test/server/Server_test.cpp [hunk 8]
```diff
@@ -367,11 +370,186 @@ class Server_test : public beast::unit_test::suite
...
+        {
+        }
+
+        ~CaptureLogs() override
+        {
+            result_ = strm_.str();
+        }
+
+        std::unique_ptr<beast::Journal::Sink>
+        makeSink(std::string const& partition,
```
Comment (mellery451) [inline src/test/server/Server_test.cpp]: yeah - it's part of the interface required of all `Logs` implementation..it's how different sinks are created for topics/partitions.

**Comment by seelabs** [comment]
In 0.80.0-b4

**Pull request closed without merging**